### PR TITLE
use twisted._threads module to address concurrency issues

### LINKED ISSUE #25: support "threadsafety=1" drivers
Trying to use Alchimia with `sqlite3`, as documented on the front page, I get:

```
	sqlalchemy.exc.ProgrammingError: (sqlite3.ProgrammingError) SQLite objects created in a thread can only be used in that same thread.The object was created in thread id 123145310715904 and this is thread id 123145306509312
```

As documented here https://www.python.org/dev/peps/pep-0249/#threadsafety in the DB-API PEP, the `threadsafety` attribute of the `sqlite3` module is `1`, which means you aren't allowed to move connections between threads.

### PR DESCRIPTION
**Pull request by glyph**
Fixes #25 

This pins each `TwistedConnection` object to a dedicated daemon thread.

### CODE REVIEW

**Comment by glyph** [comment]
NB: `twisted._threads` is nominally private, but what this means is not "don't use it", just "if you use it, you must make sure you carefully manage your Twisted upgrades in lockstep with your application" and "if it breaks during an upgrade, it's your responsibility to fix" which you should probably be doing anyway.  It was my explicit intention that other "close in" ecosystem projects experiment with it to determine if the API is ready to make public, and that's what we're doing here.

**Comment by glyph** [comment]
#26 should probably be merged first, as this includes those revisions.

**Comment by glyph** [comment]
#26 merged.

File: alchimia/engine.py [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+from twisted._threads import ThreadWorker, AlreadyQuit
+
+from twisted.internet.defer import Deferred, fail
+from twisted.python.failure import Failure
+
+from threading import Thread
+
+try:
+    from Queue import Queue
+except:
```
Comment (lvh) [inline alchimia/engine.py]: Do you really want a bare except here?

Comment (glyph) [inline alchimia/engine.py]: Yeah, good catch, I can adjust that to `except ImportError`.

File: alchimia/engine.py [hunk 2]
```diff
@@ -1,26 +1,58 @@
...
+    @worker.do
+    def container():
+        try:
+            result = work(*args, **kwargs)
+        except:
+            f = Failure()
+            deliver(lambda: deferred.errback(f))
+        else:
+            deliver(lambda: deferred.callback(result))
+    return deferred
```
Comment (lvh) [inline alchimia/engine.py]: This code feels extraordinarily generic. Why is it in alchimia?

File: alchimia/engine.py [hunk 3]
```diff
@@ -1,26 +1,58 @@
...
+                 create_worker=_threaded_worker, **kwargs):
         if reactor is None:
             raise TypeError("Must provide a reactor")
 
         self._engine = Engine(pool, dialect, url, **kwargs)
         self._reactor = reactor
-        if thread_pool is None:
-            thread_pool = reactor.getThreadPool()
-        self._tpool = thread_pool
+        self._create_worker = create_worker
```
Comment (lvh) [inline alchimia/engine.py]: The idea is that you create a new one every time you connect, right?

**Comment by lvh** [comment]
How do we document the user-facing impact of this? My understanding is that most of the APIs affected are internal, despite not being marked as such through e.g. an underscore prefix.

File: alchimia/engine.py (hunk 2 above)
Comment (glyph) [inline alchimia/engine.py]: Arguably there should be something like `deferToWorker` in `twisted._threads`, but discovering this kind of shortcoming is exactly what the "provisional" (private) status of this API is for :)

Comment (glyph) [inline alchimia/engine.py]: Probably the _really_ generic version will just take a "from" worker and a "to" worker and there will be a `_ReactorWorker` which maps `do` to `callFromThread`.

File: alchimia/engine.py (hunk 3 above)
Comment (glyph) [inline alchimia/engine.py]: Yes, and the thread lives until the end of the connection.

**Comment by glyph** [comment]
> How do we document the user-facing impact of this? My understanding is that most of the APIs affected are internal, despite not being marked as such through e.g. an underscore prefix.

#25 is the main issue in question.

One user-facing impact is that if you use a `threadsafety=1` driver, such as `sqlite3`, with the current release of alchimia, it will just start spewing exceptions at you if you try to use `engine.connect` rather than `engine.execute`; so you effectively can't use transactions.

Another user-facing impact is that if you were to try to pipeline SQL to minimize thread-communication overhead on a connection even on a `threadsafety=inf` driver, i.e. `ad = conn.execute(a); bd = conn.execute(b); cd = conn.execute(c); return gatherResults([ad, bd, cd])`, the threadpool might arbitrarily re-order the execution of `a`, `b`, and `c` despite being part of the same logical sequence.  Of course, getting this to _actually_ happen only happens under load, at scale, etc etc.

Basically, the current approach to putting work into threads is just fundamentally broken, and this fixes it.  So the main user-facing message is just "upgrade right now" :)

**Comment by lvh** [comment]
Could a user ever have passed a thread pool themselves and acquire working software at the end? That seems to be a potentially broken use case.

**Comment by glyph** [comment]
> Could a user ever have passed a thread pool themselves and acquire working software at the end? That seems to be a potentially broken use case.

Not really, no, because the operations passed to the thread pool were not tagged in any way as associated with a connection.  The breakage is intentional; hence the leading-zero version number. :).  This will probably be released as 0.7.

File: alchimia/engine.py (hunk 2 above)
Comment (lvh) [inline alchimia/engine.py]: Is there a ticket to track this?

**Comment by lvh** [comment]
Once there's a place to track the extraction of the generic bit and you fix the bare except, this looks great and I think @alex should merge it.

**Comment by glyph** [comment]
@alex gave me commit because he doesn't use this right now and has therefore approximately zero time to work on it :).  I'll merge myself once I've addressed those things.  Thanks so much for having a looK!

File: alchimia/engine.py [hunk 4]
```diff
@@ -1,26 +1,58 @@
 from __future__ import absolute_import, division
 
 from sqlalchemy.engine.base import Engine
+from sqlalchemy.exc import StatementError
 
-from twisted.internet.threads import deferToThreadPool
+from twisted._threads import ThreadWorker, AlreadyQuit
```
Comment (dreid) [inline alchimia/engine.py]: Don't use private APIs.

Comment (lvh) [inline alchimia/engine.py]: @glyph already addressed this here: https://github.com/alex/alchimia/pull/27#issuecomment-249796069

Comment (glyph) [inline alchimia/engine.py]: I am taking this comment as an additional prompt to document the intent properly on a Twisted ticket; there was some discussion in comments and on IRC but I didn't see a top-line ticket to move forward on this, so, I've filed 2: one about the API being private specifically, https://twistedmatrix.com/trac/ticket/8846, and a second about @lvh's point about the code being generic https://twistedmatrix.com/trac/ticket/8847.  Both link back to this PR.

**Comment by glyph** [comment]
All of @lvh's feedback addressed, I will land.  Thanks for looking!
